cmd/thanos/query.go

```diff
@@ -67,6 +67,8 @@ func registerQuery(m map[string]setupFunc, app *kingpin.Application, name string
 	replicaLabel := cmd.Flag("query.replica-label", "Label to treat as a replica indicator along which data is deduplicated. Still you will be able to query without deduplication using 'dedup=false' parameter.").
 		String()
 
+	defaultMaxSourceResolution := modelDuration(cmd.Flag("query.instant.default.max_source_resolution", "default value for max_source_resolution for instant queries. If not set, defaults to 0s only taking raw resolution into account.").Default("0s"))
```

Comment:
Consider improving documentation here. What are good values for this flag ? Maybe some better guide for users experiencing similiar problem to yours.

Also are you sure that this impacts only `instant` queries?

Looking at the code you changed `parseDownsamplingParamMillis` to incorporate default, which is also used in range queries.

Maybe we should have one flag for both `instant` and range queries ?

Reply:
The new function `parseDownsamplingParamMillisWithDefault` accepts the default as a parameter. For query_range that parameter is `step/5`. The `api.defaultInstantQueryMaxSourceResolution` is only used in the `query` function which corresponds to instant queries (as far as i read it).

Regarding one default for both: There has already been a default for range queries: `step/5`. As `?step` is a query parameter that is already defined by the prometheus API it provides better integration with clients than `maxSourceResolution` (at least currently). That is why i only defined the flag within the scope of instant queries. Long-Term the flag might not be necessary if we integrate `?max_source_resolution` for instant queries with clients such as grafana (or thanos' own thanos ruler).

Regarding docs: I added an example. Not sure if this bloats the documentation for a commandline flag too much though.